                                                                       TEMUULEN Bulgan - 2022427

### CCT College Dublin Continuous Assessment No.2
# AN ANALYSIS OF INDIAN FARMERS' PROTEST TWEETS.

**Brief Introduction**

1. For my second continuous assessment, I choose CSV format data of Indian Farmer's Protest Tweets. This file contains over 1 million English language tweets tweeted between November 1st, 2020 and november 21st, 2021 with the hashtag <#FarmersProtest>. It is downloaded from the Kaggle website with the CCO:Public Domain license.

2. I divided my project into 3 primary sections (Every step in data processing and analysis is fully discussed on each subsection of these primary sections.):
    1. big data storage and processing
    2. comparative analysis of databases
    3. sentiment analysis and forecast

3. I used Git for daily code tracking and GitHub for archiving, monitoring and sharing. To view the whole project on GitHub, click on the follwig link 

In [10]:
# First of all, importing libraries that I used in this project.
import pandas as pd
import os

ImportError: cannot import name 'InsecureCLient' from 'hdfs' (/home/hduser/.local/lib/python3.10/site-packages/hdfs/__init__.py)

## 1. Big Data Storage and Processing

Since the size of the data I choose is 1.8 GB it would be difficult for me to upload and process it on my VM, I decided to do some preprocessing in Python's Pandas library.

In [3]:
# Creating a variable with file size information.
file_size = os.path.getsize('/home/hduser/Desktop/ca/tweets.csv')/(1024*1024*1024)

# Printing variable value to check the size.
print("The size of the tweet file is ", round(file_size, 2), 'GB.')

The size of the tweet file is  1.69 GB.


### 1.1. Dataset prepocessing

In [4]:
# Creating a DataFrame.
df_preprocess = pd.read_csv('/home/hduser/Desktop/ca/tweets.csv')

# Checking the DataFrame.
df_preprocess

tweetUrl   
0        https://twitter.com/ShashiRajbhar6/status/1376...  \
1        https://twitter.com/kaursuk06272818/status/137...   
2        https://twitter.com/kaursuk06272818/status/137...   
3        https://twitter.com/SukhdevSingh_/status/13767...   
4        https://twitter.com/Davidmu66668113/status/137...   
...                                                    ...   
1084447  https://twitter.com/drkumarpmch2000/status/144...   
1084448  https://twitter.com/DigitalKisanBot/status/144...   
1084449  https://twitter.com/sandhumannu0476/status/144...   
1084450  https://twitter.com/mani262002/status/14433695...   
1084451  https://twitter.com/sandhumannu0476/status/144...   

                              date   
0        2021-03-30 03:33:46+00:00  \
1        2021-03-30 03:33:23+00:00   
2        2021-03-30 03:31:00+00:00   
3        2021-03-30 03:30:45+00:00   
4        2021-03-30 03:30:30+00:00   
...                            ...   
1084447  2021-09-30 00:48:05+00:00   
1084448  2021-09-30 00:28:06+00:00   
1084449  2021-09-30 00:22:20+00:00   
1084450  2021-09-30 00:18:06+00:00   
1084451  2021-09-30 00:13:02+00:00   

                                           renderedContent       tweetId   
0                             Support 👇\n\n#FarmersProtest  1.376739e+18  \
1        Supporting farmers means supporting our countr...  1.376739e+18   
2        Support farmers if you are related to food #St...  1.376739e+18   
3        #StopHateAgainstFarmers support #FarmersProtes...  1.376739e+18   
4        You hate farmers I hate you, \nif you love the...  1.376739e+18   
...                                                    ...           ...   
1084447            Good morning friends 🌹\n#FarmersProtest  1.443377e+18   
1084448  There is no compromise when it comes to injust...  1.443372e+18   
1084449  @KhalsaAid_AU @Khalsa_Aid ਮਾਨਸ ਕੀ ਜਾਤ ਸਭੈ ਏਕੈ ...  1.443371e+18   
1084450  #FarmersProtest_StandsStrong \n#FarmersProtest...  1.443370e+18   
1084451  Good morning friends... \n#FarmersProtest http...  1.443368e+18   

               userId  replyCount  retweetCount  likeCount  quoteCount   
0        1.015970e+18           0             0          0           0  \
1        1.332937e+18           0             0          0           0   
2        1.332937e+18           0             0          0           0   
3        1.308357e+18           0             1          3           0   
4        1.357312e+18           0             0          1           0   
...               ...         ...           ...        ...         ...   
1084447  3.035365e+09           1             1          1           0   
1084448  1.309303e+18           0             0          1           0   
1084449  1.377187e+18           0             0          0           0   
1084450  2.350986e+09           0             0          0           0   
1084451  1.377187e+18           0             0          0           0   

                                                    source   
0        <a href="http://twitter.com/download/android" ...  \
1        <a href="http://twitter.com/download/android" ...   
2        <a href="http://twitter.com/download/android" ...   
3        <a href="http://twitter.com/download/android" ...   
4        <a href="http://twitter.com/download/android" ...   
...                                                    ...   
1084447  <a href="http://twitter.com/download/android" ...   
1084448  <a href="https://help.twitter.com/en/using-twi...   
1084449  <a href="http://twitter.com/download/android" ...   
1084450  <a href="https://mobile.twitter.com" rel="nofo...   
1084451  <a href="http://twitter.com/download/android" ...   

                                                     media  retweetedTweet   
0                                                      NaN             NaN  \
1        [{'previewUrl': 'https://pbs.twimg.com/media/E...             NaN   
2        [{'previewUrl': 'https://pbs.twimg.com/media/E...         

In [5]:
# Examaning DataFrame.
df_preprocess.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1084452 entries, 0 to 1084451
Data columns (total 14 columns):
 #   Column           Non-Null Count    Dtype  
---  ------           --------------    -----  
 0   tweetUrl         1084452 non-null  object 
 1   date             1084452 non-null  object 
 2   renderedContent  1084452 non-null  object 
 3   tweetId          1084452 non-null  float64
 4   userId           1084452 non-null  float64
 5   replyCount       1084452 non-null  int64  
 6   retweetCount     1084452 non-null  int64  
 7   likeCount        1084452 non-null  int64  
 8   quoteCount       1084452 non-null  int64  
 9   source           1072318 non-null  object 
 10  media            319141 non-null   object 
 11  retweetedTweet   0 non-null        float64
 12  quotedTweet      286988 non-null   object 
 13  mentionedUsers   412206 non-null   object 
dtypes: float64(3), int64(4), object(7)
memory usage: 115.8+ MB


This dataset contains 8 distinct categories of information that I will use for further analysis. Rest is not important, it will use only memory and space of my VM, that's why i'm going to remove them.

In [6]:
# Deleting columns from the DataFrame
df_preprocess = df_preprocess.drop(labels=['tweetUrl', 'source', 'media', 'retweetedTweet', 'quotedTweet', 'mentionedUsers'],
                                   axis=1)

# Checking the changes.
df_preprocess.head(2)

date   
0  2021-03-30 03:33:46+00:00  \
1  2021-03-30 03:33:23+00:00   

                                     renderedContent       tweetId   
0                       Support 👇\n\n#FarmersProtest  1.376739e+18  \
1  Supporting farmers means supporting our countr...  1.376739e+18   

         userId  replyCount  retweetCount  likeCount  quoteCount  
0  1.015970e+18           0             0          0           0  
1  1.332937e+18           0             0          0           0

In [7]:
# Creating a new file of tweets on my VM with the utf-8 Encoding.
df_preprocess.to_csv('/home/hduser/Desktop/ca/new_tweets.csv', encoding='utf-8')

In [8]:
# Creating a variable with file size information.
file_size = os.path.getsize('/home/hduser/Desktop/ca/new_tweets.csv')/(1024*1024)

# Printing variable value to check the size.
print("The size of the tweet file is ", round(file_size, 2), 'MB.')

The size of the tweet file is  272.8 MB.


Now the file size is reduced to 272.8 and we still have importand information of the tweets for further processing.

### 1.2 *Storing data on distributed storage system.*

As a distributed storage system for my tweet data set I'm using Hadoop HDFS.

In [ ]:
# Establishin connection with HDFS

connection_hdfs = hdfs.HadoopFileSystem()